checker/src/org/checkerframework/checker/signedness/SignednessVisitor.java

```diff
@@ -149,6 +231,61 @@ private boolean isMaskedShift(BinaryTree shiftExpr) {
     }
 
     /**
+     * Determines if a right shift operation, {@code >>} or {@code >>>}, is type casted such that
+     * the cast renders the shift signedness ({@code >>} vs {@code >>>}) irrelevent by discarding
+     * the bits duplicated into the shift result. For example, the following pair of right shifts on
+     * {@code short s} both produce the same results under any input, because of type casting:
+     *
+     * <p>{@code (byte)(s >> 8) == (byte)(b >>> 8);}
+     *
+     * @param shiftExpr a right shift expression: {@code expr1 >> expr2} or {@code expr1 >>> expr2}
+     * @return true iff the right shift is type casted such that a signed or unsigned right shift
+     *     has the same effect
+     */
+    private boolean isCastedShift(BinaryTree shiftExpr) {
+        // enclosing is the operation or statement that immediately contains shiftExpr
+        Tree enclosing;
+        {
+            TreePath parentPath = visitorState.getPath().getParentPath();
+            enclosing = parentPath.getLeaf();
+            // Strip away all parentheses from the shift operation
+            while (enclosing.getKind() == Kind.PARENTHESIZED) {
+                parentPath = parentPath.getParentPath();
+                enclosing = parentPath.getLeaf();
+            }
+        }
+
+        if (!isPrimitiveCast(enclosing)) {
+            return false;
+        }
+
+        TypeCastTree cast = (TypeCastTree) enclosing;
+
+        // Determine the type of the cast target
```

Comment:
This code feels repetitive to me.  `isPrimitiveCast` just did much of this work, and now you are doing it again.
How about replacing `isPrimitiveCast` by `primitiveCastType` that returns null if the argument is not a primitive cast, and otherwise returns the type?  Then the client (this code) can just test against null.  It reduces code duplication, reduces run time, and makes this code shorter and easier to understand by abstracting out implementation details.